scalalib/test/src/mill/scalalib/scalafmt/ScalafmtTests.scala

```diff
@@ -0,0 +1,105 @@
+package mill.scalalib.scalafmt
+
+import ammonite.ops._
+import mill.main.Tasks
+import mill.scalalib.ScalaModule
+import mill.util.{TestEvaluator, TestUtil}
+import utest._
+import utest.framework.TestPath
+
+object ScalafmtTests extends TestSuite {
+
+  trait TestBase extends TestUtil.BaseModule {
+    def millSourcePath =
+      TestUtil.getSrcPathBase() / millOuterCtx.enclosing.split('.')
+  }
+
+  object ScalafmtTestModule extends TestBase {
+    object core extends ScalaModule with ScalafmtModule {
+      def scalaVersion = "2.12.4"
+    }
+  }
+
+  val resourcePath = pwd / 'scalalib / 'test / 'resources / 'scalafmt
+
+  def workspaceTest[T, M <: TestUtil.BaseModule](
+      m: M,
+      resourcePath: Path = resourcePath)(t: TestEvaluator[M] => T)(
+      implicit tp: TestPath): T = {
+    val eval = new TestEvaluator(m)
+    rm(m.millSourcePath)
+    rm(eval.outPath)
+    mkdir(m.millSourcePath / up)
+    cp(resourcePath, m.millSourcePath)
+    t(eval)
+  }
+
+  def tests: Tests = Tests {
+    'scalafmt - {
+      def checkReformat(reformatCommand: mill.define.Command[Unit]) =
+        workspaceTest(ScalafmtTestModule) { eval =>
+          val before = getProjectFiles(ScalafmtTestModule.core, eval)
+
+          // first reformat
+          val Right(_) = eval.apply(reformatCommand)
+
+          val firstReformat = getProjectFiles(ScalafmtTestModule.core, eval)
+
+          assert(
+            firstReformat("Main.scala").modifyTime > before("Main.scala").modifyTime,
+            firstReformat("Main.scala").content != before("Main.scala").content,
+            firstReformat("Person.scala").modifyTime > before("Person.scala").modifyTime,
+            firstReformat("Person.scala").content != before("Person.scala").content,
+            // resources files aren't modified
```

Comment:
I wanted to separate different parts of the test with nested test cases(like 'firstTime, 'cached, 'afterChange), but it didn't work, because I needed shared test evaluator, and if I did nested tests - test evaluator was created every time for each test. I couldn't create `TestEvaluator` outside of the `Tests` block either because it requires `implicit TestPath`, that is not available outside of the `Tests` block.
So, is it possible to do this with `utest`?

Reply:
It's not easy to do with uTest, leaving them in the same test seems ok for now